objects/oWeakling/Draw_0.gml

```diff
@@ -19,22 +25,53 @@ if instance_exists(_weakling_inst) {
 	}
 }
 
+// Sync oWeakling's sprite animation frame with the progress bar. This allows
+// the animation to speed up as we purchase speed upgrades. 
+// TODO: How can we write this to account for possible range-of-motion upgrades?
 // progress bar instance ID
-_inst = inst_61B8692F
-
-// TODO: Generalize this so that it works for more than one frame, so we
-// can offer an upgrade to range-of-motion. 
+var _inst = inst_61B8692F
 if instance_exists(_inst) {
+
+	var _range_of_motion = 4
+	// If sprite_get_number returns 2, our frames are 0 and 1. 	
+	// Pick the smaller value - total number of frames in the sprite, or total upgradeable
+	//  range of motion. 
+	var _number_of_frames = min(sprite_get_number(oWeakling.sprite_index), _range_of_motion)
+	
+	// In order to account for reversed frames, double the number of frames and
+	// subtract 2 (one for the beginning of the animation, one for the end). 
+	var _true_number_of_frames = 2 * _number_of_frames -  2
 	
-	// number_of_frames = array_length(image_index)
+	// Figure out how far along the progress bar is. 
+	var _progress_percent = (_inst.alarm[0] / (global.repSpeed * room_speed)) * 100
 	
-	// if the progress bar is halfway complete, set this animation frame
-	if (((_inst.alarm[0] / (global.repSpeed * room_speed)) * 100) <= 50) {
-		image_index = 0
+	// What is the division of 100% that we need? For 2 frames, this will be 50%, 
+	// for three frames, this will be 33%, etc. 
+	var _original_progress_increment = floor(100 / _true_number_of_frames)
 	
-	// otherwise, set that animation frame
-	} else {
-		image_index = 1
+	var _image_index_value = 0
+	var _progress_increment = _original_progress_increment
+	while(true) {
+		
+		// If the progress bar is less than the current calculated increment,
+		// this is where we want to set the image! 
+		if _progress_percent < _progress_increment {
+			image_index = _image_index_value
+			break
+		
+		// Otherwise, advance the window that we're checking, and the image 
+		// index we'll use once we find a match. 
+		} else {
+			// If we're past the 50% mark, progress the animation.
+			if _progress_increment < 50 {
```

Comment:
I think this is the culprit - it should be less than _or equal to_ 50%. For example, for two frames, we should progress the animation at 50%, otherwise we'll roll back.

Reply:
Fixed in https://github.com/ggrammar/IdleLearning3/pull/6/commits/cce6b9dfb28423be80c0ad50e2e54274fdc286f4